### PR TITLE
with Stripe Connect a user might have several customers

### PR DESCRIPTION
**Pull request by ticosax**
#### What's this PR do?
Add an option to prevent customer to be linked to a user. For stripe Connect support

#### Definition of Done (check if considered and/or addressed):

- [x] Are all backwards incompatible changes documented in this PR?
- [x] Have all new dependencies been documented in this PR?
- [x] Has the appropriate documentation been updated (if applicable)?
- [x] Have you written tests to prove this change works (if applicable)?

### CODE REVIEW

File: pinax/stripe/actions/customers.py [hunk 1]
```diff
@@ -35,6 +35,9 @@ def create(user, card=None, plan=settings.PINAX_STRIPE_DEFAULT_PLAN, charge_imme
         charge_immediately: whether or not the user should be immediately
                             charged for the subscription
         quantity: the quantity (multiplier) of the subscription
+        stripe_account: An account object. If given, the Customer and User relation will not be established for you.
+                        The caller is responsible to maintain relationship between the User and the Customer
+                        in the context of each Account. Because a single User might have several Customers, one for each Account.
```
Comment (blueyed) [inline pinax/stripe/actions/customers.py]: s/one for each/one per account/ ?!

File: pinax/stripe/actions/customers.py [hunk 2]
```diff
@@ -47,11 +50,16 @@ def create(user, card=None, plan=settings.PINAX_STRIPE_DEFAULT_PLAN, charge_imme
         quantity=quantity,
         trial_end=trial_end
     )
+    if stripe_account is None:
+        kwargs = {"user": user}
+    else:
+        # we want to allow several customer per user, each for every stripe account
```
Comment (blueyed) [inline pinax/stripe/actions/customers.py]: s/customer/customers/
s/each for every/for any/

**Comment by paltman** [comment]
Looks good to me but I'm admittedly not up to speed on Connect yet.  @lukeburden thoughts?

**Comment by lukeburden** [comment]
@paltman Basically, yes - Connect accounts can have their own customer entries. We will need a one-to-many between the User model and pinax.stripe.Customer.

@ticosax Your changes allow creation of Customer instances without association with any User. It isn't providing for multiple Customer entries per User, so far as I can tell. How would you look up relevant Customer entries if the `user` field is None?

I think to properly implement this, we'd need:

- an N-to-1 between Customer and User
- to consider the backwards compatibility ramifications for the `customer` reverse to the 1-to-1 that currently is projected onto the User model by the Customer model. What happens when the User has multiple Customers?
- we'd need to change Customer to inherit from `AccountRelatedStripeObject` such that Customer instances have a `stripe_account` field.
- in the `customers.create` method and when syncing customer data from webhooks, we'd need to ensure that the `stripe_account` is populated
- we'd need a lookup method in `actions.customers.py` to lookup a Customer for a given User instance with an optional `stripe_account` parameter, so you can select the right Customer instance if there are multiple.

Fair bit of work, sadly. :/ I didn't need Customers within connected accounts so didn't come across this. Sorry!

**Comment by ticosax** [comment]
Thank you @lukeburden, I agree this PR was too premature. I will continue to make some progress and come back with a proposal addressing the point you mentioned. stay tuned.

**Comment by ticosax** [comment]
depends on #402

**Comment by blueyed** [comment]
This will be replaced with a UserAccount through model (https://github.com/pinax/pinax-stripe/pull/440/files#diff-90b0e5e37b96af44d789c587800e86c0R234).

**Comment by paltman** [comment]
@ticosax @blueyed if this will be replaced, then is it worth keeping open?

**Comment by blueyed** [comment]
@paltman 
Please close it, I will create a new PR - but likely not anymore today.

**Pull request closed without merging**